google-cloud-bigquery/src/main/java/com/google/cloud/bigquery/BigQueryImpl.java

```diff
@@ -468,46 +446,72 @@ public TableDataInsertAllResponse call() throws Exception {
+  @Override
+  public TableResult listTableData(
+      String datasetId, String tableId, Schema schema, TableDataListOption... options) {
+    return listTableData(TableId.of(datasetId, tableId), schema, options);
+  }
+
+  @Override
+  public TableResult listTableData(TableId tableId, Schema schema, TableDataListOption... options) {
+    Tuple<? extends Page<FieldValueList>, Long> data =
+        listTableData(tableId, getOptions(), optionMap(options));
+    return new TableResult(schema, data.y(), data.x());
+  }
+
+  private static Tuple<? extends Page<FieldValueList>, Long> listTableData(
+      final TableId tableId,
+      final BigQueryOptions serviceOptions,
+      final Map<BigQueryRpc.Option, ?> optionsMap) {
     try {
       final TableId completeTableId = tableId.setProjectId(serviceOptions.getProjectId());
-      Tuple<String, Iterable<TableRow>> result =
-          runWithRetries(new Callable<Tuple<String, Iterable<TableRow>>>() {
-            @Override
-            public Tuple<String, Iterable<TableRow>> call() {
-              return serviceOptions.getBigQueryRpcV2()
-                  .listTableData(completeTableId.getProject(), completeTableId.getDataset(),
-                      completeTableId.getTable(), optionsMap);
-            }
-          }, serviceOptions.getRetrySettings(), EXCEPTION_HANDLER, serviceOptions.getClock());
-      String cursor = result.x();
-      return new PageImpl<>(new TableDataPageFetcher(tableId, serviceOptions, cursor, optionsMap),
-          cursor, transformTableData(result.y(), null));
+      TableDataList result =
+          runWithRetries(
+              new Callable<TableDataList>() {
+                @Override
+                public TableDataList call() {
+                  return serviceOptions
+                      .getBigQueryRpcV2()
+                      .listTableData(
+                          completeTableId.getProject(),
+                          completeTableId.getDataset(),
+                          completeTableId.getTable(),
+                          optionsMap);
+                }
+              },
+              serviceOptions.getRetrySettings(),
+              EXCEPTION_HANDLER,
+              serviceOptions.getClock());
+      String cursor = result.getPageToken();
+      return Tuple.of(
+          new PageImpl<>(
+              new TableDataPageFetcher(tableId, serviceOptions, cursor, optionsMap),
+              cursor,
+              transformTableData(result.getRows())),
+          result.getTotalRows());
     } catch (RetryHelper.RetryHelperException e) {
       throw BigQueryException.translateAndThrow(e);
     }
   }
 
-  private static List<FieldValueList> transformTableData(
-      Iterable<TableRow> tableDataPb, Schema schema) {
-    final FieldList schemaFields = schema != null ? schema.getFields() : null;
+  private static Iterable<FieldValueList> transformTableData(Iterable<TableRow> tableDataPb) {
```

Comment:
Do you recall why the schema argument was dropped from this method?

Reply:
It was weirdly cumbersome to add it here. The responsibility of adding schema is giving to `TableResult` instead.